crates/libm-test/src/special_case.rs

```diff
@@ -0,0 +1,231 @@
+        None
+    }
+}
+
+impl MaybeOverride<(f32,)> for SpecialCase {
+    fn check_float<F: Float>(
+        input: (f32,),
+        actual: F,
+        expected: F,
+        _ulp: &mut u32,
+        ctx: &CheckCtx,
+    ) -> Option<TestResult> {
+        if ctx.basis == CheckBasis::Musl {
+            if ctx.fname == "acoshf" && input.0 < -1.0 {
+                // acoshf is undefined for x <= 1.0, but we return a random result at lower
+                // values.
+                return XFAIL;
+            }
+
+            if ctx.fname == "sincosf" {
+                let factor_frac_pi_2 = input.0.abs() / f32::consts::FRAC_PI_2;
+                if (factor_frac_pi_2 - factor_frac_pi_2.round()).abs() < 1e-2 {
+                    // mwe have a bad approximation near multiples of pi/2
+                    return XFAIL;
+                }
+            }
+
+            if ctx.fname == "expm1f" && input.0 > 80.0 && actual.is_infinite() {
+                // we return infinity but the number is representable
+                return XFAIL;
+            }
+
+            if ctx.fname == "sinhf" && input.0.abs() > 80.0 && actual.is_nan() {
+                // we return some NaN that should be real values or infinite
+                // doesn't seem to happen on x86
+                return XFAIL;
+            }
+
+            if ctx.fname == "lgammaf" || ctx.fname == "lgammaf_r" && input.0 < 0.0 {
+                // loggamma should not be defined for x < 0, yet we both return results
+                return XFAIL;
+            }
+        }
+
+        maybe_check_nan_bits(actual, expected, ctx)
+    }
+}
+
+impl MaybeOverride<(f64,)> for SpecialCase {
+    fn check_float<F: Float>(
+        input: (f64,),
+        actual: F,
+        expected: F,
+        _ulp: &mut u32,
+        ctx: &CheckCtx,
+    ) -> Option<TestResult> {
+        if ctx.basis == CheckBasis::Musl {
+            if cfg!(target_arch = "x86") && ctx.fname == "acosh" && input.0 < 1.0 {
+                // The function is undefined, both implementations return random results
+                return SKIP;
+            }
+
+            if cfg!(x86_no_sse)
+                && ctx.fname == "ceil"
+                && input.0 < 0.0
+                && input.0 > -1.0
+                && expected == F::ZERO
+                && actual == F::ZERO
+            {
+                // musl returns -0.0, we return +0.0
+                return XFAIL;
+            }
+
+            if ctx.fname == "lgamma" || ctx.fname == "lgamma_r" && input.0 < 0.0 {
+                // loggamma should not be defined for x < 0, yet we both return results
+                return XFAIL;
+            }
+        }
+
+        maybe_check_nan_bits(actual, expected, ctx)
+    }
+}
+
+/// Check NaN bits if the function requires it
+fn maybe_check_nan_bits<F: Float>(actual: F, expected: F, ctx: &CheckCtx) -> Option<TestResult> {
+    if !(ctx.canonical_name == "abs" || ctx.canonical_name == "copysigh") {
```

Comment:
Another great catch. I plan to change everything to enum matching at some point to avoid these kind of issues, but that will be a follow up since this PR already has a lot going on.